cohttp-eio/src/client.ml

```diff
@@ -0,0 +1,127 @@
+module Buf_read = Eio.Buf_read
+module Buf_write = Eio.Buf_write
+
+type response = Http.Response.t * Buf_read.t
+type host = string * int
+type scheme = [ `HTTP | `HTTPS ]
+type 'a conn = scheme -> host -> Eio.Flow.two_way as 'a
+
+type 'a body_disallowed_call =
+  ?version:Http.Version.t ->
+  ?headers:Http.Header.t ->
+  'a conn ->
+  Uri.t ->
+  response
+(** [body_disallowed_call] denotes HTTP client calls where a request is not
+    allowed to have a request body. *)
+
+type 'a body_allowed_call =
+  ?version:Http.Version.t ->
+  ?headers:Http.Header.t ->
+  ?body:Body.t ->
+  'a conn ->
+  Uri.t ->
+  response
+
+(* Request line https://datatracker.ietf.org/doc/html/rfc7230#section-3.1.1 *)
+let write_request writer (meth, version, headers, resource_path, body) =
+  Buf_write.string writer (Http.Method.to_string meth);
+  Buf_write.char writer ' ';
+  Buf_write.string writer resource_path;
+  Buf_write.char writer ' ';
+  Buf_write.string writer (Http.Version.to_string version);
+  Buf_write.string writer "\r\n";
+  Rwer.write_headers writer headers;
+  Buf_write.string writer "\r\n";
+  Body.write_body writer body
+
+(* response parser *)
+
+let is_digit = function '0' .. '9' -> true | _ -> false
+
+let status_code =
+  let open Rwer in
+  let open Buf_read.Syntax in
+  let+ status = take_while1 is_digit in
+  Http.Status.of_int (int_of_string status)
+
+let reason_phrase =
+  Buf_read.take_while (function
+    | '\x21' .. '\x7E' | '\t' | ' ' -> true
+    | _ -> false)
+
+(* https://datatracker.ietf.org/doc/html/rfc7230#section-3.1.2 *)
+let response buf_read =
+  let open Buf_read.Syntax in
+  match Buf_read.at_end_of_input buf_read with
+  | true -> Stdlib.raise_notrace End_of_file
+  | false ->
+      let version = Rwer.(version <* space) buf_read in
+      let status = Rwer.(status_code <* space) buf_read in
+      let () = Rwer.(reason_phrase *> crlf *> return ()) buf_read in
+      let headers = Rwer.http_headers buf_read in
+      Http.Response.make ~version ~status ~headers ()
+
+(* Generic HTTP call *)
+
+let call ?(meth = `GET) ?(version = `HTTP_1_1) ?(headers = Http.Header.init ())
+    ?(body = Body.Empty) flow_fn uri =
+  let host_name =
+    match Uri.host uri with
+    | Some h -> h
+    | None -> raise (invalid_arg "uri: host missing")
+  in
+  let scheme =
+    match Uri.scheme uri with
+    | Some "http" -> `HTTP
+    | Some "https" -> `HTTPS
+    | Some _ -> raise (invalid_arg "uri: invalid uri scheme")
```

Comment:
In the current cohttp clients, you can override the resolvers to for example do http over sockets, see e.g. https://github.com/mirage/ocaml-cohttp#creating-custom-resolver-a-docker-socket-client-example can you do this here?

Reply:
Do you specifically mean connecting to a UNIX domain socket? Otherwise the example client and tests already provides a demo of how to create a connection. In any case I think connecting to a domain socket would a good test case so I will add it to the test case.

Reply:
Yes, I meant unix socket. Two common questions I had in the past are how to timeout requests and how to connect to unix domain sockets. Since with ocaml5+eio it is done differently, I think we should document it either in the interface, in an example to link to the README or in the README, so that we avoid answering the same questions all the time.

Reply:
Added docker_client.ml which mimics the cohttp-lwt-unix example.

Reply:
> Two common questions I had in the past are how to timeout requests and how to connect to unix domain sockets. Since with ocaml5+eio it is done differently, I think we should document it either in the interface, in an example to link to the README or in the README, so that we avoid answering the same questions all the time.

Regarding timeouts, `Eio.Time.with_timeout` should do it. However, an example might be nice, indeed.